### PR TITLE
appModuleHandler.AppModule.appArchitecture: detect AMD64 apps in Windows 11 on ARM via kernel32.dll::GetProcessInformation

### LINKED ISSUE #14403: Windows 11 on ARM: "ARM64" is returned for app architecture for AMD64 applications
Hi,

Stemming from a discussion in #14397:

### Background:
#7894 introduced a collection of properties for app modules, one of them being app architecture. The purpose of "app architecture" property is to inform humans and code about which machine architecture the app is designed for. This property internally asks IsWow64Process2 function to return the machine type the process is running under, which is used to figure out if a process is a WoW64 process (x86 on x64, for example). There is a design flaw: in Windows 11 on ARM, it is possible to run x64 apps, causing NVDA to think that this is an ARM64 process when it is not.

### Steps to reproduce:
See #14397 for an example, but to recap: run AMD64 app on a Windows 11 on ARM machine and observe that, in NVDA Python Console, typing "focus.apModule.appArchitecture" returns "ARM64" when it should say "AMD64".

### Actual behavior:

"ARM64" is returned as ap architecture property.

### Expected behavior:

"AMD64" is returned.

#### NVDA installed/portable/running from source:
Installed/portable

#### NVDA version:
2022.3.2

#### Windows version:
Windows 11 22H2 (build 22623 beta)

#### Name and version of other software in use when reproducing the issue:
None

#### Does the issue still occur after restarting your computer?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
None

#### If NVDA add-ons are disabled, is your problem still occurring?
Yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

### Solution:
In Windows 11 and later, ask kernel32.dll::GetProcessInformation to obtain actual process machine information (although GetProcessInformation is available in Windows 8 and later, Windows 11 is required since one of the parameters was made available with build 22000).

Thanks.

### PR DESCRIPTION
**Pull request by josephsl**
### Link to issue number:
Closes #14403 
Stems from #14397 

### Summary of the issue:
In Windows 11 on ARM, AMD64 apps are seen as ARM64 apps.

### Description of user facing changes
Entering "focus.appModule.appArchitecture" in Python Console will say "AMD64" if focused on AMD64 application.

### Description of development approach
In app module handler:

1. Defined a process information structure with three attributes: process machine (unsigned short), res0 (double word), machine attributes (double word.
2. Added 0xAA64 to represent ARM64 applications in app architecture property.
3. Refined app architecture property in AppModule class to call kernel32.dll::GetProcessInformation on Windows 11 21H2 (build 22000) or later, with WoW64 detection reserved for Windows 10 and earlier.
4. If GetProcessInformation fails (returns 0), report "unknown" as app architecture to prepare to handle cases where an app (perhaps hosted or an ARM64EC) may report wrong architecture information.
5. Added type information to app architecture property (return type: str). This should inspire efforts to add type information to app module handler odule and classes (later).

### Testing strategy:
Manual testing (ARM64 machine is ideal but also works on x64 machines):

1. If not done so, upgrade to Windows 11 on a supported ARM-based computer (Surface Pro X, for example).
2. Install and run an application known to be designed for AMD64 such as Notepad++ or Slack.
3. While focused on the app, press Control+NVDA+Z to open Python Console and type "focus.appModule.appArchitecture" and observe the value (expected: AMD64).

### Known issues with pull request:
None, although it might be possible that some x64 apps may continue to be seen as ARM64, in which case the development approach may need to be refined.

### Change log entries:
Bug fixes (and changes for developers):

In Windows 11 on ARM, x64 apps are no longer identified as ARM64 applications. (#14403)

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by rperez030** [comment]
Tested using a Windows on ARM VM. X86, AMD64 and ARM64 where reported correctly for Slack (X86), Remote Incident Manager (AMD64) and Visual Studio code (ARM64), however, when tested with 1Password 8 which should have been reported as AMD64, NVDA generated the following error:

Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "baseObject.pyc", line 26, in __get__
  File "appModuleHandler.pyc", line 710, in _get_appArchitecture
KeyError: 0

NVDA 2022.3.2 incorrectly reports ARM64 for the same app.

**Comment by josephsl** [comment]
Hi, sounds like it is saying “machine unknown” for this particular app. We might as well say “ARM64” for this case unless this is truly an AMD64 app and/or is an EC app. CC @MichaelDCurran both for your comment and a review please.

**Comment by rperez030** [comment]
I believe it is an AMD64, or maybe EC. I'm going by their website where they indicate that it requires Windows 11 64-bit or Windows 10 64-bit

**Comment by leonardder** [comment]
I'm indeed pretty sure that 1Password is AMD64 only.

File: source/appModuleHandler.py [hunk 1]
```diff
@@ -682,28 +690,40 @@ def _get_appArchitecture(self):
...
-				self.appArchitecture = archValues2ArchNames[processMachine.value]
-		except AttributeError:
-			# Windows 10 Version 1507 (build 10240) and earlier.
-			self.appArchitecture = "AMD64" if self.is64BitProcess else "x86"
+		# #14403: GetProcessInformation can be called from Windows 11 and later to obtain process machine.
+		if winVersion.getWinVer() >= winVersion.WIN11:
+			processMachineInfo = _PROCESS_MACHINE_INFORMATION()
+			# Constant comes from PROCESS_INFORMATION_CLASS enumeration.
+			ProcessMachineTypeInfo = 9
+			ctypes.windll.kernel32.GetProcessInformation(
```
Comment (michaelDCurran) [inline source/appModuleHandler.py:704]: Definitely should check the return value here. If false then we should report as unknown.
My hypothesis is that for 1Password  this function fails due to process handle permissions.
However, if this is not the case, then at very least, using `get` or catching KeyError on the arch dictionary will obviously be necessary.

File: source/appModuleHandler.py [hunk 2]
```diff
@@ -684,28 +692,43 @@ def _get_appArchitecture(self):
...
+			# Constant comes from PROCESS_INFORMATION_CLASS enumeration.
+			ProcessMachineTypeInfo = 9
+			# Sometimes getProcessInformation may fail, so say "unknown".
+			if not ctypes.windll.kernel32.GetProcessInformation(
+				self.processHandle,
+				ProcessMachineTypeInfo,
+				ctypes.byref(processMachineInfo),
+				ctypes.sizeof(_PROCESS_MACHINE_INFORMATION)
+			):
+				self.appArchitecture = "unknown"
```
Comment (michaelDCurran) [inline source/appModuleHandler.py:713]:
```suggestion
				log.debugWarning("Could not query process architecture")
				self.appArchitecture = "unknown"
```

File: source/appModuleHandler.py [hunk 3]
```diff
@@ -684,28 +692,43 @@ def _get_appArchitecture(self):
...
 			else:
-				# On ARM64, two 32-bit architectures are supported: x86 (via emulation) and ARM (natively).
-				self.appArchitecture = archValues2ArchNames[processMachine.value]
-		except AttributeError:
-			# Windows 10 Version 1507 (build 10240) and earlier.
-			self.appArchitecture = "AMD64" if self.is64BitProcess else "x86"
+				self.appArchitecture = archValues2ArchNames.get(processMachineInfo.ProcessMachine, "unknown")
```
Comment (michaelDCurran) [inline source/appModuleHandler.py:715]: Add debug warnings for both situations. 
```suggestion
				try:
					self.appArchitecture = archValues2ArchNames[processMachineInfo.ProcessMachine]
				except KeyError:
					log.debugWarning(f"Unknown process architecture 0X{processMachineInfo.ProcessMachine:X}")
					self.appArchitecture = "unknown"
```

**Review by michaelDCurran — CHANGES_REQUESTED** [review]

**Review by michaelDCurran — APPROVED** [review]

**Comment by michaelDCurran** [comment]
Having tested this on ARM64 now:
The appArchitecture of 1Password is failing because getProcessInformation returns 0, because the appModule's processHandle is actually 0. I.e. NVDA was unable to OpenProcess on 1Password.
However, the processHandle property of any NVDAObject from 1Password is valid. Oleacc's GetProcessHandleFromHwnd is working.
Therefore, it might be an idea to update an appModule's processHandle property if it is 0, with an NvDAObject's valid one.
I would suggest doing this in `appModuleHandler.GetAppModuleForNVDAObject`.
Btw, 1Password is in deed AMD64 on ARM64.

**Comment by michaelDCurran** [comment]
It is also worth noting that in the 1password case, the appModule is also staying alive for ever, as passing a processHandle of 0 to WaitForSingleObject is returning a True value. (WAIT_FAIL). We should consider improving the isAlive property to only return True for WAIT_TIMEOUT, False for WAIT_0, and anything else should be logged as n error and probably return False.

**Comment by josephsl** [comment]
Hi,

I'll incorporate Mick's suggestion this week after merging latest master commit on AMD64 support on Windows 11 ARM64.

Thanks.

**Comment by josephsl** [comment]
Hi,

Perhaps a follow-up work: when running unit tests, process handle is undefined for placeholder NVDA object. For now NVDA will ignore it via attribute error.

Thanks.

**Comment by josephsl** [comment]
Hi,

Pull request build: https://ci.appveyor.com/api/buildjobs/x6a2kw2muycblho7/artifacts/output%2Fnvda_snapshot_pr14404-27340%2C5a08f6a6.exe

While running this build, see if 1password is recognized as an AMD64 application under ARM64 (open NVDA Pyhton Console, type focus.appModule.appArchitecture).

Thanks.

**Review by seanbudd — APPROVED** [review]

**Pull request merged**